### ISSUE TITLE
[Bug]: Axes placement and tight_layout()

### ISSUE THREAD

**Issue by suuuehgi** [matplotlib/matplotlib#24226]
### Bug summary

`Figure.tight_layout()` changes relative placement of axes.

I don't know where this is a bug, a feature request or just unavoidable behavior.

(In this simple case, moving `.tight_layout()` ahead of `.add_axes` fixes it.)

### Code for reproduction

```python
from matplotlib import pyplot as plt
import matplotlib

fig, axs = plt.subplots( nrows=2, sharex=True )

# left, bottom, width, height
box = matplotlib.transforms.Bbox.from_bounds(.5, 0, .1, .1)

fig.add_axes(fig.transFigure.inverted().transform_bbox( axs[0].transAxes.transform_bbox(box)))
fig.add_axes(fig.transFigure.inverted().transform_bbox( axs[1].transAxes.transform_bbox(box)))

fig.tight_layout()
```

### Actual outcome

![github](https://user-images.githubusercontent.com/7861774/196914057-89377adc-f6b4-4806-a2a3-c7ab1ac47247.jpg)

### Expected outcome

![github2](https://user-images.githubusercontent.com/7861774/196914113-6b85fb88-7dd9-4e91-8c0b-3c5e6d4e121e.jpg)

### Additional information

Related issue: #24225

### Operating system

Arch

### Matplotlib Version

3.6.1

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.10.7

### Jupyter version

Jupyter Lab v. 3.4.5

### Installation

pip

**Comment by jklymak** [comment]
This is expected behaviour.  Have you considered using ax.inset_axes for your use case?

**Comment by suuuehgi** [comment]
> This is expected behaviour. Have you considered using ax.inset_axes for your use case?

Eureka! 🥳 Didn't know that method existed! Thank you!

For anyone stumbling upon this in the future:

```python
fig, axs = plt.subplots( nrows=2, sharex=True )

# left, bottom, width, height
axs[0].inset_axes( [.5, 0, .1, .1] )
axs[1].inset_axes( [.5, 0, .1, .1] )

fig.tight_layout()
```

**Comment by jklymak** [comment]
https://matplotlib.org/stable/gallery/subplots_axes_and_figures/colorbar_placement.html is meant to answer that question.  Let us know if that tutorial doesn't make sense.  You may also prefer to use `layout='constrained'` in most cases to tight_layout.

**Comment by suuuehgi** [comment]
> https://matplotlib.org/stable/gallery/subplots_axes_and_figures/colorbar_placement.html is meant to answer that question. Let us know if that tutorial doesn't make sense. You may also prefer to use `layout='constrained'` in most cases to tight_layout.

Thank you very much for the hint! I did not know about `constrained_layout`! Afaics the `layout` argument is not mentioned in the docstring of `subplots`. I updated my answer above.

Regarding the tutorial: The tutorial is fine but it falls short on colorbars "on top of / within plots" as sketched above.

For matrix plotting, like `imshow`, `pcolormesh` and alike, the colorbar naturally goes next to the plot as written within the tutorial.
But for function plotting / `scatter`-pots, it might be interesting to move them inwards as I did above.
